snapcraft/extractors/_metadata.py

```diff
@@ -15,27 +15,37 @@
 # along with this program.  If not, see <http://www.gnu.org/licenses/>.
 
 import yaml
-from typing import Any, Dict
+from typing import Any, Dict, List
 
 
 class ExtractedMetadata(yaml.YAMLObject):
     """Collection of metadata extracted from a part."""
 
     yaml_tag = u'!ExtractedMetadata'
 
-    def __init__(self, *, summary='', description='') -> None:
+    def __init__(
+            self, *, summary: str='',
+            description: str='', icon: str='',
+            desktop: List[str]=None) -> None:
         """Create a new ExtractedMetadata instance.
 
         :param str summary: Extracted summary
         :param str description: Extracted description
-        """
+        :param str icon: Extracted icon
+        :param dict desktop: Extracted desktop ids, as defined in
+            https://standards.freedesktop.org/desktop-entry-spec/desktop-entry-spec-latest.html#desktop-file-id
+        """  # noqa
 
-        self._data = {}  # type: Dict[str, str]
+        self._data = {}  # type: Dict[str, Any]
```

Comment:
I tried with Union, but then it required me to use as return type for all the other attributes Union[str, List[str]] instead of just str. That felt wrong, because then if we add a new type we would have to modify all the return types. Apparently, it is not smart enough. Or maybe I was doing something wrong, have you tried it?

Reply:
Well, that's correct, is it not? Any value that comes out of this dict could be either a string or a list of strings.

Reply:
I think I got it with your hint about casting. Please take a look at the last commit.